### PR TITLE
[Linter] Correcting rule to constrain for Azure management plane only

### PR DESCRIPTION
**Pull request by vishrutshah**
fixes https://github.com/Azure/autorest/issues/1938

### CODE REVIEW

**Comment by dsgouda** [comment]
@olydis thoughts?

**Comment by vishrutshah** [comment]
Closing the issue based on offline talk with @dsgouda https://github.com/Azure/autorest/issues/1940

**Comment by vishrutshah** [comment]
@azuresdkci Test this please

**Comment by vishrutshah** [comment]
@azuresdkci Test this please

**Comment by vishrutshah** [comment]
@azuresdkci Test this please

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 1]
```diff
@@ -328,7 +328,7 @@ Modelers:
     <value>Parameters "subscriptionId" and "api-version" are not allowed in the operations section</value>
   </data>
   <data name="ServiceDefinitionParametersMissingMessage" xml:space="preserve">
-    <value>Parameters "subscriptionId" or "api-version" are referenced but not defined in the parameters section of Service Definition</value>
+    <value>Parameters "subscriptionId" and "api-version" must be defined in the parameters section in the swagger for the Azure Management Plane service.</value>
```
Comment (dsgouda) [inline src/core/AutoRest.Core/Properties/Resources.resx]: 1. Update this message
2. There is a different rule that flags a case where subscriptionId and ApiVersion are not defined in the global parameters section, you might want to modify the message to reflect that these parameters are not expected in the operations' parameters section

File: src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs [hunk 2]
```diff
@@ -33,7 +37,27 @@ public class ServiceDefinitionParameters : TypedRule<Dictionary<string, SwaggerP
...
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">Formatted parameters</param>
+        /// <returns></returns>
+        public override bool IsValid(Dictionary<string, SwaggerParameter> ParametersMap, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new object[0];
+            ServiceDefinition serviceDefinition = context.Root;
+            
+            // Rule does not apply if x-ms-parameterized-host extension is not applied
+            if (!serviceDefinition.Extensions.ContainsKey(XMsParameterizedHostExtensionName))
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs]: If we are not checking the management host, this becomes redundant, just simplify this to checking operation parameters instead

**Comment by dsgouda** [comment]
Please check you have the tests updated to reflect this rule

File: src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs [hunk 3]
```diff
@@ -33,7 +37,27 @@ public class ServiceDefinitionParameters : TypedRule<Dictionary<string, SwaggerP
...
+            {
+                return true;
+            }
 
+            return true == (ParametersMap?.Values.Any(parameter => parameter.Name == SubscriptionId || parameter.Name == ApiVersion));
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs]: hmm... totally missed this, should be case insensitive right?

**Comment by veronicagg** [comment]
@vishrutshah looks like CI is failing and branch has some conflicts.

File: src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs (hunk 3 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs]: I think this should be case-sensitive. That'd also facilitate us towards common definitions. Ref: https://github.com/Azure/azure-rest-api-specs-pr/pull/40/files#diff-d3a0dcdc5d2939703fa9e32dff2295c1R249

File: src/core/AutoRest.Core/Properties/Resources.resx (hunk 1 above)
Comment (vishrutshah) [inline src/core/AutoRest.Core/Properties/Resources.resx]: Could you please point me to the rule that validates that they must be in global parameter section? 

Based on rule id [M2014](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) seems like this rule supposed to do that but current implementation doesn't differentiate.

File: src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs (hunk 3 above)
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs]: Agreed, but existing specs may not be following any convention

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 4]
```diff
@@ -325,7 +325,7 @@ Modelers:
     <value>Parameters "subscriptionId" and "api-version" are not allowed in the operations section, define these in the global parameters section instead</value>
   </data>
   <data name="ServiceDefinitionParametersMissingMessage" xml:space="preserve">
-    <value>Parameter "{0}" is referenced but not defined in the global parameters section of Service Definition</value>
+    <value>Parameters "subscriptionId" and "api-version" must be defined in the parameters section in the swagger for the Azure Management Plane service.</value>
```
Comment (veronicagg) [inline src/core/AutoRest.Core/Properties/Resources.resx]: minor: I think we could also say for "Azure Resource Management service", not sure what's best, so feel free to pick

File: src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs (hunk 3 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs]: I'd like to catch those getting caught with these rule then.

File: src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs [hunk 5]
```diff
@@ -50,7 +36,21 @@ public override IEnumerable<ValidationMessage> GetValidationMessages(Dictionary<
...
+        /// <param name="ParametersMap">Dictionary of swagger parameters</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">Formatted parameters</param>
+        /// <returns></returns>
+        public override bool IsValid(Dictionary<string, SwaggerParameter> ParametersMap, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new object[0];
+            ServiceDefinition serviceDefinition = context.Root;
+
+            return true == (ParametersMap?.Values.Any(parameter => parameter.Name == SubscriptionId || parameter.Name == ApiVersion));
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ServiceDefinitionParameters.cs]: Definitely think this check should be case insensitive

**Pull request closed without merging**